jdaviz/configs/default/plugins/data_quality/data_quality.py

```diff
@@ -0,0 +1,271 @@
+
+        with delay_callback(dq_layer.state, 'alpha', 'cmap', 'v_min', 'v_max'):
+            if len(flag_bits):
+                dq_layer.state.v_min = min(flag_bits)
+                dq_layer.state.v_max = max(flag_bits)
+
+            dq_layer.state.alpha = self.dq_layer_opacity
+            dq_layer.state.cmap = cmap
+
+    def get_dq_layer(self):
+        if self.dq_layer_selected == '':
+            return
+
+        viewer = self.viewer.selected_obj
+        [dq_layer] = [
+            layer for layer in viewer.layers if
+            layer.layer.label == self.dq_layer_selected
+        ]
+        return dq_layer
+
+    def get_science_layer(self):
+        viewer = self.viewer.selected_obj
+        [science_layer] = [
+            layer for layer in viewer.layers if
+            layer.layer.label == self.science_layer_selected
+        ]
+        return science_layer
+
+    @observe('dq_layer_opacity')
+    def update_opacity(self, event={}):
+        science_layer = self.get_science_layer()
+        dq_layer = self.get_dq_layer()
+
+        if dq_layer is not None:
+            # DQ opacity is a fraction of the science layer's opacity:
+            dq_layer.state.alpha = self.dq_layer_opacity * science_layer.state.alpha
+
+    @observe('decoded_flags', 'flags_filter')
+    def update_cmap(self, event={}):
+        dq_layer = self.get_dq_layer()
+        flag_bits = np.array([flag['flag'] for flag in self.decoded_flags])
+        rgb_colors = [hex2color(flag['color']) for flag in self.decoded_flags]
+
+        hidden_flags = np.array([
+            flag['flag'] for flag in self.decoded_flags
+
+            # hide the flag if the visibility toggle is False:
+            if not flag['show'] or
+
+            # hide the flag if `flags_filter` has entries but not this one:
+            (
+                len(self.flags_filter) and
+                not np.isin(list(flag['decomposed'].keys()), list(self.flags_filter)).any()
+            )
+        ])
+
+        with delay_callback(dq_layer.state, 'v_min', 'v_max', 'alpha', 'stretch', 'cmap'):
+            # set correct stretch and limits:
+            # dq_layer.state.stretch = 'lookup'
+            stretch_object = dq_layer.state.stretch_object
+            stretch_object.flags = flag_bits
+            stretch_object.dq_array = dq_layer.get_image_data()
+            stretch_object.hidden_flags = hidden_flags
+
+            # update the colors of the listed colormap without
+            # reassigning the layer.state.cmap object
+            cmap = dq_layer.state.cmap
+            cmap.colors = rgb_colors
+            cmap._init()
+
+            # trigger updates to cmap in viewer:
+            dq_layer.update()
+
+            if len(flag_bits):
+                dq_layer.state.v_min = min(flag_bits)
+                dq_layer.state.v_max = max(flag_bits)
+
+            dq_layer.state.alpha = self.dq_layer_opacity
+
+    def update_visibility(self, index):
```

Comment:
is index (here and in `update_color`) convenient for the user?  This isn't a big deal since we're not exposing this in the user API yet, but definitely something to think about as we play with this in workflows, etc.

Reply:
Good point. Right now, the index corresponds to an entry in `self.decoded_flags`, which is a _list_ of dictionaries. We could instead choose to make `decoded_flags` a dictionary, where the key is the flag. I made it a list because I had better luck iterating over a list in the vue file than a dict.

Reply:
Right, but even if it internally is a dictionary, maybe we should support passing the key and just _internally_ iterating to look up the entry to change?